### PR TITLE
[SEDONA-199] Added ST_NDims to Apache Sedona

### PR DESCRIPTION
**Pull request by Amrit-Bhaskar-abhask10**
## Did you read the Contributor Guide?

- Yes, I have read [Contributor Rules](https://sedona.apache.org/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/community/develop/)

## Is this PR related to a JIRA ticket?

- Yes, the URL of the associated JIRA ticket is https://issues.apache.org/jira/projects/SEDONA/issues/SEDONA-199. The PR name follows the format `[SEDONA-199] my subject`.

## What changes were proposed in this PR?
Added ST_NDims functions to SQL

## How was this patch tested?
Through unit tests written in Scala

## Did this PR include necessary documentation updates?
Yes, I have updated the documentation.

### CODE REVIEW

File: common/src/main/java/org/apache/sedona/common/Functions.java [hunk 1]
```diff
@@ -280,6 +280,19 @@ public static int nPoints(Geometry geometry) {
         return geometry.getNumPoints();
     }
 
+    public static int nDims(Geometry geometry) {
+        String str = geometry.getCoordinate().toString();
```
Comment (jiayuasu) [inline common/src/main/java/org/apache/sedona/common/Functions.java:284]: Why not directly check NaN using getX getY getZ getM from coordinate class?

File: docs/api/sql/Function.md [hunk 2]
```diff
@@ -1475,3 +1475,24 @@ SELECT ST_ZMin(ST_GeomFromText('LINESTRING(1 3 4, 5 6 7)'))
 ```
 
 Output: `4.0`
+
+## ST_NDims
+
```
Comment (jiayuasu) [inline docs/api/sql/Function.md:1480]: 1. Please add more elaboration and examples about dimension when the input has M or Z.
2. The output should be `3` integer not `3.0` double. Please correct me if I am wrong.
3. The doc should be sorted by the alphabetical order

File: sql/src/test/scala/org/apache/sedona/sql/functionTestScala.scala [hunk 3]
```diff
@@ -459,6 +459,11 @@ class functionTestScala extends TestBaseScala with Matchers with GeometrySample
 
     }
 
+    it("Passed ST_NDims") {
```
Comment (jiayuasu) [inline sql/src/test/scala/org/apache/sedona/sql/functionTestScala.scala:462]: Please add more test cases for 2, 3, 4D (e.g., when Z and M present, or both present)

**Review by jiayuasu — CHANGES_REQUESTED** [review]
Thanks for your contribution. Please make changes accordingly.

**Comment by Amrit-Bhaskar-abhask10** [comment]
Hi @jiayuasu, The requested changes are fixed now. Please let us know if any more changes are required.

File: sql/src/test/scala/org/apache/sedona/sql/functionTestScala.scala [hunk 4]
```diff
@@ -459,6 +459,31 @@ class functionTestScala extends TestBaseScala with Matchers with GeometrySample
...
+      assert(test.take(1)(0).get(0).asInstanceOf[Int] == 2)
+    }
+
+    it("Passed ST_NDims with Z coordinates") {
+      val test = sparkSession.sql("SELECT ST_NDims(ST_GeomFromWKT('POINTZ(1 1 0.5)'))")
+      assert(test.take(1)(0).get(0).asInstanceOf[Int] == 3)
+    }
+
+    it("Passed ST_NDims with both Z and M coordinates") {
+      val test = sparkSession.sql("SELECT ST_NDims(ST_GeomFromWKT('POINTZM(1 1 1 0.5)'))")
```
Comment (jiayuasu) [inline sql/src/test/scala/org/apache/sedona/sql/functionTestScala.scala:483]: According to PostGIS doc (https://postgis.net/docs/ST_NDims.html), XYZM should return 4, but your case returns 3. If you cannot support M, I am ok with this. But please clearly state it in the API doc.

**Review by jiayuasu — CHANGES_REQUESTED** [review]

File: sql/src/test/scala/org/apache/sedona/sql/functionTestScala.scala (hunk 4 above)
Comment (Kontinuation) [reply sql/src/test/scala/org/apache/sedona/sql/functionTestScala.scala:483]: Currently the geometry serializer in `sedona-sql` does not support M dimension, 4D geometries with ZM coordinates will have their M coordinates dropped and became 3D geometries. We're working on [a new geometry serializer](https://issues.apache.org/jira/projects/SEDONA/issues/SEDONA-207?filter=allopenissues) to resolve this issue.

Comment (Amrit-Bhaskar-abhask10) [reply sql/src/test/scala/org/apache/sedona/sql/functionTestScala.scala]: Thanks, @Kontinuation for clarifying my doubt.

Comment (Amrit-Bhaskar-abhask10) [reply sql/src/test/scala/org/apache/sedona/sql/functionTestScala.scala]: @jiayuasu, I made the updates with the test cases and documentation.

**Review by jiayuasu — APPROVED** [review]

**Pull request merged**